% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/datasets.R
\docType{data}
\name{pf.annot}
\alias{pf.annot}
\title{All \emph{P. falciparum} gene-product annotations.}
\format{
A data frame of 5720 rows and 9 columns.

\describe{
\item{seqid}{chromosome ID.}
\item{geneStart}{gene start coordinate.}
\item{geneEnd}{gene end coordinate.}
\item{strand}{gene strand.}
\item{type}{type of annotation. Here protein-coding genes, ncRNA genes and pseudogenes are kept.}
\item{geneID}{gene ID.}
\item{description}{functional annotation.}
\item{geneName}{the gene symbol, or short name, if one exists; otherwise the geneID is used.}
\item{source}{annotation source.}
}
}
\source{
\url{https://plasmodb.org/common/downloads/release-68/Pfalciparum3D7/gff/data/PlasmoDB-68_Pfalciparum3D7.gff}
}
\usage{
data(pf.annot)
}
\description{
A dataset containing all \emph{P. falciparum} gene-product annotations taken directly from gff file(reference strain 3D7). Only "gene" entries are kept to remove redundancy.
}
\details{
\subsection{Keep in mind:}{

These data aren't explicitly required for running any enrichments with this package. They are included for reference to aid exploring your enrichment results.

Some redundant columns were filtered/removed from the original data source.

Updated versions can be generated using the \link{get.pfannot} function and specifying the url or filepath to your preferred plasmoDB .gff file. Note the run.topGO.meta function currently calls get.pfannot with plasmoDB release 68 to annotate the significant-genes-in-enriched-terms output.
}
}
\keyword{dataset}
