% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/Utilities.R
\name{run.topGO.meta}
\alias{run.topGO.meta}
\title{Run JO's topGO pipeline}
\usage{
run.topGO.meta(
  mydf = "mydf",
  geneID2GO = "Pfal_geneID2GO_curated",
  pval = 0.05
)
}
\arguments{
\item{mydf}{data frame with geneIDs in column 1, and interest-category classifications in column 2.}

\item{geneID2GO}{A list of named vectors of GO IDs--one vector of GO-terms for each geneID.}

\item{pval}{pvalue threshold for significance. Defaults to 0.05.}
}
\description{
Tests for functional enrichment in gene-categories of interest.
}
\details{
The \strong{run.topGO.meta} function:
\itemize{
\item defines which genes are "interesting" and which should be defined as background for each category specified in mydf,
\item makes the GOdata object for topGO,
\item tests each category of interest for enriched GO-terms against all the other genes included in mydf (the "gene universe"),
\item and then outputs results to table (.tsv files that can be opened in Excel).
}

Enrichments are performed by each ontology (molecular function, biological process, cellular compartment) sequentially on all groups of interest. Results are combined in the final output-table ("Routput/GO/all.combined.GO.results.tsv").

TopGO automatically accounts for genes that cannot be mapped to GO terms (or are mapped to terms with < 3 genes in the analysis) with "feasible genes" indicated in the topGO.log files in the "Routput/GO" folder.
}
\section{\strong{outputs}}{

run.topGO.meta creates several output-files, including:
\itemize{
\item enrichment results,
\item significant genes per significant term,
\item plots of the GO-term hierarchy relevant to the analysis, and
\item thorough log-files for each gene-category of interest tested against the background of all other genes in the analysis.
}

Primary results from run.topGO.meta will be in "Routput/GO/all.combined.GO.results.tsv".
}

\section{\strong{Concepts for common use-cases}}{


\emph{RNAseq}:
In an RNAseq analysis, common categories might be "upregulated", "downregulated", and "neutral". The gene universe would consist of all genes detected above your threshold cutoffs (\emph{not necessarily all genes in the genome}).

\emph{piggyBac screens}:
In pooled \emph{piggyBac}-mutant screening, common categories might be "sensitive", "tolerant", and "neutral". The gene universe would consist of all genes represented in your screened library of mutants (\emph{again, not all genes in the genome}).
See the included \link{exampleMydf} as an example.
}

\section{\strong{Using your own custom GO database}}{


A correctly formatted geneID2GO object is included for P. falciparum enrichment analyses (\link{Pfal_geneID2GO}). You may also provide your own, so long as it is a named character-vector of GO-terms (each vector named by geneID, with GO terms as each element).

You can use the included \code{\link[=formatGOdb.curated]{formatGOdb.curated()}} function to format a custom GO database from curated GeneDB/PlasmoDB annotations for several non-model organisms (or the \code{\link[=formatGOdb]{formatGOdb()}} function to include all GO annotations, if you aren't picky about including automated electronic annotations). If you're studying a model organism, several annotations are already available and can be downloaded through the AnnotationDbi bioconductor package that loads with topGO.
}

\examples{

data(exampleMydf)  ## load exampleMydf object (mutant classification-data from pooled 1k heat shock screen)
data(Pfal_geneID2GO_curated)  ## load Pfal_geneID2GO_curated object
run.topGO.meta(exampleMydf,Pfal_geneID2GO_curated)

}
\seealso{
\code{\link[topGO:topGO-package]{topGO::topGO()}}
}
